File: Algorithms/p086_Partition_List/p086_Partition_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        lt = ListNode(None)
        gt_eq = ListNode(None)
        p1 = lt
        p2 = gt_eq
        while head:
            if head.val<x:
                p1.next = head
                p1 = head
            else:
                p2.next = head
                p2 = head
            head = head.next
        p1.next = gt_eq.next
        p2.next = None
        return lt.next if lt.next else gt_eq.next

File: Algorithms/p086_Partition_List/p086_Partition_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode lt(0);
        ListNode gt_eq(0);
        ListNode* p1 = &lt;
        ListNode* p2 = &gt_eq;
        while(head){
            if(head->val<x){
                p1->next = head;
                p1 = head;
            }else{
                p2->next = head;
                p2 = head;
            }
            head = head->next;
        }
        p1->next = gt_eq.next;
        p2->next = 0;
        return lt.next ? lt.next : gt_eq.next;
    }
};
